src/frontend/Typechecker.ml

```diff
@@ -0,0 +1,1520 @@
+    | Error (ty, meta) ->
+        Semantic_error.mismatched_array_types meta.loc ty meta.type_ |> error
+    | Ok (ad_level, type_) ->
+        let type_ = UnsizedType.UArray type_ in
+        mk_typed_expression ~expr:(ArrayExpr es) ~ad_level ~type_ ~loc )
+
+let check_rowvector loc es =
+  match es with
+  | {emeta= {ad_level; type_= UnsizedType.URowVector; _}; _} :: elements -> (
+    match get_consistent_types ad_level URowVector elements with
+    | Ok (ad_level, _) ->
+        mk_typed_expression ~expr:(RowVectorExpr es) ~ad_level ~type_:UMatrix
+          ~loc
+    | Error (_, meta) ->
+        Semantic_error.invalid_matrix_types meta.loc meta.type_ |> error )
+  | _ -> (
+    match get_consistent_types DataOnly UReal es with
+    | Ok (ad_level, _) ->
+        mk_typed_expression ~expr:(RowVectorExpr es) ~ad_level
+          ~type_:URowVector ~loc
+    | Error (_, meta) ->
+        Semantic_error.invalid_row_vector_types meta.loc meta.type_ |> error )
+
+(* index checking *)
+
+let indexing_type idx =
+  match idx with
+  | Single {emeta= {type_= UnsizedType.UInt; _}; _} -> `Single
+  | _ -> `Multi
+
+let inferred_unsizedtype_of_indexed ~loc ut indices =
+  let rec aux type_ idcs =
+    match (type_, idcs) with
+    | _, [] -> type_
+    | UnsizedType.UArray type_, `Single :: tl -> aux type_ tl
+    | UArray type_, `Multi :: tl -> aux type_ tl |> UnsizedType.UArray
+    | (UVector | URowVector), [`Single] | UMatrix, [`Single; `Single] ->
+        UnsizedType.UReal
+    | (UVector | URowVector | UMatrix), [`Multi] | UMatrix, [`Multi; `Multi] ->
+        type_
+    | UMatrix, ([`Single] | [`Single; `Multi]) -> UnsizedType.URowVector
+    | UMatrix, [`Multi; `Single] -> UnsizedType.UVector
+    | UMatrix, _ :: _ :: _ :: _
+     |(UVector | URowVector), _ :: _ :: _
+     |(UInt | UReal | UComplex | UFun _ | UMathLibraryFunction), _ :: _ ->
+        Semantic_error.not_indexable loc ut (List.length indices) |> error
+  in
+  aux ut (List.map ~f:indexing_type indices)
+
+let inferred_ad_type_of_indexed at uindices =
+  UnsizedType.lub_ad_type
+    ( at
+    :: List.map
+         ~f:(function
+           | All -> UnsizedType.DataOnly
+           | Single ue1 | Upfrom ue1 | Downfrom ue1 ->
+               UnsizedType.lub_ad_type [at; ue1.emeta.ad_level]
+           | Between (ue1, ue2) ->
+               UnsizedType.lub_ad_type
+                 [at; ue1.emeta.ad_level; ue2.emeta.ad_level])
+         uindices )
+
+(* function checking *)
+
+let verify_conddist_name loc id =
+  if
+    List.exists
+      ~f:(fun x -> String.is_suffix id.name ~suffix:x)
+      Utils.conditioning_suffices
+  then ()
+  else Semantic_error.conditional_notation_not_allowed loc |> error
+
+let verify_fn_conditioning loc id =
+  if
+    List.exists
+      ~f:(fun suffix -> String.is_suffix id.name ~suffix)
+      Utils.conditioning_suffices
+    && not (String.is_suffix id.name ~suffix:"_cdf")
+  then Semantic_error.conditioning_required loc |> error
+  else ()
```

Comment:
We've been trying to move away from all this `String.is_suffix` business. The function type contains a field with type `Fun_kind.suffix` and these suffix checks should use that whenever possible.